runtime/src/main/java/org/corfudb/runtime/collections/streaming/StreamingTask.java

```diff
@@ -149,14 +177,53 @@ private Optional<CorfuStreamEntries> transform(ILogData logData) {
         return Optional.of(new CorfuStreamEntries(streamEntries, timestamp));
     }
 
-    public DeltaStream getStream() {
-        return this.stream;
+    public List<DeltaStream> getStreamsList() {
+        return this.streamsList;
+    }
+
+    /* This function either returns null when none of the streams have any new addresses to read
+     * or returns a stream
+     *  (i) when either of the streams have a trimGap(). This is because trimGap() means there is
+     * data that has not been synced.
+     *  (ii) when either of the streams have data. If a subset of streams have data, then the stream
+     *  with the lowest address is returned
+     */
+    private DeltaStream findNextStreamToRead() {
+        DeltaStream nextStream = null;
+        long smallestAddress = Address.MAX;
+
+        for (DeltaStream stream : streamsList) {
+            if (!stream.hasNext()) {
+                continue;
+            }
+            if (stream.trimGap()) {
+                nextStream = stream;
+               break;
+            }
+            if (nextStream == null) {
+                nextStream = stream;
+            }
+
+            long nextAddress = stream.getNextAddressToRead();
+            if (smallestAddress > nextAddress) {
+                nextStream = stream;
+                smallestAddress = nextAddress;
+            }
+        }
+
+        return nextStream;
     }
 
     private void produce() {
         Preconditions.checkState(status.get() == StreamStatus.SYNCING);
-        Preconditions.checkState(stream.hasNext());
-        ILogData logData = stream.next();
+        currentStream = findNextStreamToRead();
+        // if neither of the streams have updates, move the task to runnable
+        if (currentStream == null) {
+            move(StreamStatus.SYNCING, StreamStatus.RUNNABLE);
+            return;
+        }
```

Comment:
How can it arrive to this branch?

Since produce will only be called if hasNext returns true, when produce is called it either needs to produce actual entries for the listener or put the stream in an error state (i.e., encountering an Exception)